packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
+    // can start outside of the frame's time boundaries and could pose a risk
+    // for us missing a frame.
 
-      if (_asyncEvents[asyncUID] == null) {
-        frames[event.frameId].addEvent(current);
-      }
+    // Ensure the event fits within the frame's time boundaries.
+    if (!_eventOccursWithinFrameBoundaries(e, f)) return false;
+
+    bool eventAdded = false;
+
+    if (e.isPipelineProduceFlow && f.pipelineProduceFlow == null) {
+      f.pipelineProduceFlow = e;
+      eventAdded = true;
+    } else if (e.isPipelineConsumeFlow && f.pipelineConsumeFlow == null) {
+      f.pipelineConsumeFlow = e;
+      eventAdded = true;
     }
+
+    // Adding event 'e' could mean we have completed the frame. Check if we
+    // should add the completed frame to [_frameCompleteController].
+    _maybeAddCompletedFrame(f);
+
+    return eventAdded;
   }
 
-  TimelineEventType _getTimelineEventType(TraceEvent event) {
-    // The event will only be of type 'cpu' or 'gpu' because we do not process
-    // events of any other type.
-    if (event.isCpuEvent) {
-      return TimelineEventType.cpu;
-    } else {
-      return TimelineEventType.gpu;
+  bool _eventOccursWithinFrameBoundaries(TimelineEvent e, TimelineFrame f) {
+    // Epsilon in microseconds.
+    const int epsilon = 50;
+
+    // Allow the event to extend the frame boundaries by `epsilon` microseconds.
+    final bool fitsStartBoundary = f.startTime - e.startTime - epsilon < 0;
+    final bool fitsEndBoundary = f.endTime - e.endTime + epsilon > 0;
+    return fitsStartBoundary && fitsEndBoundary;
+  }
+
+  List<TimelineFrame> _getAndSortWellFormedFrames() {
+    final List<TimelineFrame> frames =
+        List<TimelineFrame>.from(_pendingFrames.values)
+            .where((TimelineFrame f) => f.wellFormed)
+            .toList();
+
+    // Sort frames by their startTime. Sorting these frames ensures we will
+    // handle the oldest frame first when iterating through the list.
+    frames.sort((a, b) => a.startTime.compareTo(b.startTime));
+    return frames;
+  }
+
+  void _maybeAddCompletedFrame(TimelineFrame frame) {
+    if (frame.readyForTimeline && !frame.addedToTimeline.isCompleted) {
+      _frameCompleteController.add(frame);
+      _pendingFrames.remove(frame);
+      frame.addedToTimeline.complete();
     }
   }
 }
 
+/// Data describing a single frame.
+///
+/// Each TimelineFrame should have 4 distinct pieces of data:
+///   - [_pipelineProduceFlow] : flow of events showing the CPU work for the
+///     frame.
+///   - [_pipelineConsumeFlow] : flow of events showing the GPU work for the
+///     frame.
 class TimelineFrame {
-  TimelineFrame(this.id, this.startTime);
+  TimelineFrame(this.id);
 
   final String id;
 
-  final List<TimelineEvent> cpuEvents = [];
-  final List<TimelineEvent> gpuEvents = [];
+  /// Marks whether this frame has been added to the timeline.
+  Completer<Null> addedToTimeline = Completer();
```

Comment:
generally make Completer objects private and only expose Future objects in the API

Reply:
Hmm I need to complete this from outside the class.
```
    if (frame.readyForTimeline && !frame.addedToTimeline.isCompleted) {
      ...
      frame.addedToTimeline.complete();
    }
```

I could either 
a) make the completer private, add a public method to complete it, and add a boolean getter to return addedToTimeline.isCompleted
or b) accomplish the same end with a public boolean. 

I liked the completer because it can't be reset once it is marked complete (the original code should have been final - will fix if I keep it).

Thoughts?